server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
+                                    ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
+                                ),
+                                nodeConnectionHistory.disconnectCause
+                            );
+                        } else {
+                            logger.warn(
+                                """
+                                    reopened transport connection to node [{}] \
+                                    which disconnected gracefully [{}ms] ago but did not \
+                                    restart, so the disconnection is unexpected; \
+                                    if unexpected, see [{}] for troubleshooting guidance""",
+                                node.descriptionWithoutAttributes(),
+                                nodeConnectionHistory.disconnectTime,
+                                ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
+                            );
+                        }
+                    }
+                }
+
+                @Override
+                public void onNodeDisconnected(DiscoveryNode node, Transport.Connection connection) {
+                    connection.addCloseListener(new ActionListener<Void>() {
+                        @Override
+                        public void onResponse(Void ignored) {
+                            insertNodeConnectionHistory(null);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            insertNodeConnectionHistory(e);
+                        }
+
+                        private void insertNodeConnectionHistory(@Nullable Exception e) {
+                            final long disconnectTime = threadPool.absoluteTimeInMillis();
+                            final NodeConnectionHistory nodeConnectionHistory = new NodeConnectionHistory(
+                                node.getEphemeralId(),
+                                disconnectTime,
+                                e
+                            );
+                            final String nodeId = node.getId();
+                            NodeConnectionHistory previousConnectionHistory = nodeHistory.get(nodeId);
+                            if (previousConnectionHistory != null) {
+                                nodeHistory.replace(nodeId, previousConnectionHistory, nodeConnectionHistory);
+                            }
+                        }
+                    });
+                }
+            });
+        }
+
+        void reserveConnectionHistoryForNodes(DiscoveryNodes nodes) {
+            for (DiscoveryNode node : nodes) {
+                nodeHistory.put(node.getId(), dummy);
+            }
+        }
+
+        void removeConnectionHistoryForNodes(Set<DiscoveryNode> nodes) {
+            final int startSize = nodeHistory.size();
+            for (DiscoveryNode node : nodes) {
+                nodeHistory.remove(node.getId());
```

Comment:
There's kind of an implicit invariant here that `org.elasticsearch.cluster.NodeConnectionsService.ConnectionHistory#nodeHistory` and `org.elasticsearch.cluster.NodeConnectionsService#targetsByNode` have the same keys. At the very least we should be able to assert this. I also wonder if we should be calling `nodeHistory.retainAll()` to make it super-clear that we are keeping these keysets aligned.

But then that got me thinking, maybe we should be tracking the connection history of each target node in `ConnectionTarget` rather than trying to maintain two parallel maps. Could that work?